Source/NSPathControl.m

```diff
@@ -0,0 +1,396 @@
+  
+  result = [op runModalForDirectory: nil
+                               file: nil
+                              types: nil];
+  if (result == NSOKButton)
+    {
+      NSArray *urls = [op URLs];
+      NSURL *url = [urls objectAtIndex: 0];
+      [self setURL: url];
+    }
+
+  [[sender menu] close];
+}
+
+- (void) mouseDown: (NSEvent *)event
+{
+  if (![self isEnabled])
+    {
+      [super mouseDown: event];
+      return;
+    }
+
+  if ([self pathStyle] == NSPathStylePopUp)
+    {
+      NSPathCell *acell = (NSPathCell *)[self cell];
+      NSArray *array = [acell pathComponentCells];
+      NSMenu *menu = [[NSMenu alloc] initWithTitle: @"Select File"];
+      NSPathComponentCell *c = nil;
+      NSEnumerator *en = [array objectEnumerator];
+      
+      while((c = [en nextObject]) != nil)
+        {
+          NSURL *u = [c URL];
+          NSString *s = [[u path] lastPathComponent];
+          NSMenuItem *i = [[NSMenuItem alloc] init];
+
+          [i setTitle: s];
+          [i setTarget: self];
+          [i setAction: @selector(_doMenuAction:)];
+          
+          [menu insertItem: i
+                   atIndex: 0]; 
+        }
+
+      // Add separator
+      [menu insertItem: [NSMenuItem separatorItem]
+               atIndex: 0];
+
+      // Add choose menu option
+      NSMenuItem *i = [[NSMenuItem alloc] init];
+      [i setTitle: @"Choose..."];
+      [i setTarget: self];
+      [i setAction: @selector(_doChooseMenuAction:)];
+      [menu insertItem: i
+               atIndex: 0];
+      
+      if (_delegate)
+        {
+          if ([(id)_delegate respondsToSelector: @selector(pathControl:willPopUpMenu:)])
+            {
+              [_delegate pathControl: self
+                       willPopUpMenu: menu];
+            }
+        }
+
+      [menu popUpMenuPositionItem: [menu itemAtIndex: 0]
+                       atLocation: NSMakePoint(0.0, 0.0)
+                           inView: self];
+    }
+  else
+    {
+      if (_action)
+        {
+          [self sendAction: _action
+                        to: _target];
+        }
+    }
+}
+
+- (NSDragOperation) draggingEntered: (id<NSDraggingInfo>)sender
+{
+  // if (_delegate != nil)
+    {
+      NSDragOperation d = [_delegate pathControl: self
+                                    validateDrop: sender];
+      if (d == NSDragOperationCopy)
+        {
+          NSPasteboard *pb = [sender draggingPasteboard];
+          if ([[pb types] containsObject: NSFilenamesPboardType])
+            {
+              NSArray *files = [pb propertyListForType: NSFilenamesPboardType];
+              if ([files count] > 0)
+                {
+                  NSString *file = [files objectAtIndex: 0];
+                  NSURL *u = [NSURL URLWithString: file];
+                  BOOL accept = [_delegate pathControl: self
+                                            acceptDrop: sender];
+                  if (accept)
+                    {
+                      [self setURL: u];
```

Comment:
This should only be called in performDragOperation:
Please have a look at the implementation in NSImageView.